### PR TITLE
Dynamic reload of SSL certificates for NGINX Plus

### PR DESCRIPTION
**Pull request by oseoin**
### Proposed changes

Enable [lazy loading of SSL Certificates](https://www.nginx.com/blog/nginx-plus-r18-released/#dynamic-certificate-loading_disk) for NGINX Plus. This works by replacing the `/etc/nginx/secrets` with a `map` which returns that path by default for the variable `$secret_dir_path`.

E.g.
```
ssl_certificate /etc/nginx/secrets/default-my-secret.crt
```

Becomes:
```
map $nginx_version $secret_dir_path {
    default "/etc/nginx/secrets";
}

...

ssl_certificate $secret_dir_path/default-my-secret.crt
```

When this is enabled NIC checks secret updates and skips a reload of NGINX if there are no changes to the configuration files which reference that secret.

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

File: charts/nginx-ingress/values.yaml [hunk 1]
```diff
@@ -463,6 +463,9 @@ controller:
   ## Configure root filesystem as read-only and add volumes for temporary data.
   readOnlyRootFilesystem: false
 
+  ## Enable dynamic reloading of certificates for NGINX Plus
+  enableSSLDynamicReload: true
```
Comment (vepatel) [inline charts/nginx-ingress/values.yaml:467]: new flag is required to be added in chart README.md

**Comment by vepatel** [comment]
new cli arg needs to be added in https://github.com/nginxinc/kubernetes-ingress/blob/dynamic-reload-ssl/docs/content/configuration/global-configuration/command-line-arguments.md

File: tests/suite/test_tls.py [hunk 2]
```diff
@@ -131,3 +144,41 @@ def test_tls_termination(self, kube_apis, ingress_controller_endpoint, test_name
         replace_secret(kube_apis.v1, tls_setup.secret_name, test_namespace, tls_setup.new_secret_path)
         wait_before_test(1)
         assert_gb_subject(ingress_controller_endpoint, tls_setup.ingress_host)
+
+        count_after = get_reload_count(tls_setup.metrics_url)
+        reloads = count_after - count_before
+        expected_reloads = 8
```
Comment (vepatel) [inline tests/suite/test_tls.py:150]: this count will always be deterministic?

File: cmd/nginx-ingress/flags.go [hunk 3]
```diff
@@ -269,6 +275,11 @@ func parseFlags() {
 	if *ingressLink != "" && *externalService != "" {
 		glog.Fatal("ingresslink and external-service cannot both be set")
 	}
+
+	if *enableDynamicSSLReload && !*nginxPlus {
+		glog.V(3).Infof("%s flag requires -nginx-plus and will not be enabled", dynamicSSLReloadParam)
```
Comment (vepatel) [inline cmd/nginx-ingress/flags.go:280]: should this be fatal? similar to maybe:
`*enableAppProtect && !*nginxPlus`

Comment (oseoin) [reply cmd/nginx-ingress/flags.go:280]: Since it is opt-out rather than opt-in we would fail for OSS with fatal, so I think it should stay info?

**Comment by vepatel** [comment]
https://github.com/nginxinc/kubernetes-ingress/actions/runs/7170099267/job/19524954747?pr=4764#step:3:1741 seems to have some assertions based on ssl certs resulting in failure

**Review by pdabelf5 — APPROVED** [review]

File: tests/suite/test_tls.py (hunk 2 above)
Comment (oseoin) [reply tests/suite/test_tls.py]: This has changed in the latest updates, it now only expects a 1 or 0 depending on is dynamic loads is enabled.

File: tests/suite/test_transport_server_tcp_load_balance.py [hunk 4]
```diff
@@ -617,5 +623,88 @@ def test_secure_tcp_request_load_balanced(
...
+        print(f"replacing: {sec_name} in {transport_server_setup.namespace}")
+        replace_secret(kube_apis.v1, sec_name, transport_server_setup.namespace, src_new_sec_yaml)
+        wait_before_test()
+        print(f"waited to {datetime.now().strftime('%m/%d/%Y, %H:%M:%S')}")
+        count_after = get_reload_count(transport_server_setup.metrics_url)
+        reloads = count_after - count_before_replace
+        expected_reloads = 0
+        assert reloads == expected_reloads, f"expected {expected_reloads} reloads, got {reloads}"
+
+        # self.restore_ts(kube_apis, transport_server_setup)
```
Comment (vepatel) [inline tests/suite/test_transport_server_tcp_load_balance.py:709]: might need to uncomment this line to reset TS back to its original state

File: tests/suite/test_transport_server_tcp_load_balance.py [hunk 5]
```diff
@@ -617,5 +623,88 @@ def test_secure_tcp_request_load_balanced(
...
+                    "-v=3",
+                ],
+            },
+            {"example": "transport-server-tcp-load-balance"},
+        )
+    ],
+    indirect=True,
+)
+class TestTransportServerTcpLoadBalanceDynamicReload:
+    def test_secure_tcp_request_load_balanced(
```
Comment (vepatel) [inline tests/suite/test_transport_server_tcp_load_balance.py:664]: I think we need to change the test name here to better reflect what's being tested, something like:
`test_dynamic_reload()` should be sufficient

File: tests/suite/test_transport_server_tcp_load_balance.py [hunk 6]
```diff
@@ -617,5 +623,88 @@ def test_secure_tcp_request_load_balanced(
...
+        self, kube_apis, crd_ingress_controller, transport_server_setup, ingress_controller_prerequisites
+    ):
+        """
+        Sends requests to a TLS enabled load balanced TCP service.
```
Comment (vepatel) [inline tests/suite/test_transport_server_tcp_load_balance.py:668]: need to update the description inline with test name.

**Comment by oseoin** [comment]
> ```
>   File "/workspace/tests/suite/test_virtual_server_tls.py", line 201, in test_tls_termination
>     assert reloads == expected_reloads, f"expected {expected_reloads} reloads, got {reloads}"
> AssertionError: expected 0 reloads, got 1
> assert 1 == 0
> ```

Should hopefully be resolved in the latest commit, always passed locally but upstream ordering for the virtualserver resource was not stable before now.

**Review by shaun-nx — APPROVED** [review]
👏

**Review by vepatel — APPROVED** [review]

**Pull request merged**